### PR TITLE
MM-47111 : Migrate "components/emoji_picker/emoji_picker_tabs" to typescript

### PR DESCRIPTION
**Pull request by alexpjohnson**
#### Summary

Migrates components/emoji_picker/emoji_picker_tabs to Typescript.
#### Ticket Link

  Fixes https://github.com/mattermost/mattermost-server/issues/21059
Jira: https://mattermost.atlassian.net/browse/MM-47111
#### Release Note

```release-note

```

### CODE REVIEW

File: components/emoji_picker/emoji_picker_tabs.tsx [hunk 1]
```diff
@@ -11,29 +10,42 @@ import {makeAsyncComponent} from 'components/async_load';
...
+    onEmojiClick: (emoji: Emoji) => void;
+    onGifClick?: (gif: string) => void;
+    enableGifPicker?: boolean;
+}
+
+type State = {
+    emojiTabVisible: boolean;
+    filter: string;
+}
+export default class EmojiPickerTabs extends PureComponent<Props, State> {
```
Comment (M-ZubairAhmed) [inline components/emoji_picker/emoji_picker_tabs.tsx:41]: nit : empty line in between

File: components/emoji_picker/emoji_picker_tabs.tsx [hunk 2]
```diff
@@ -11,29 +10,42 @@ import {makeAsyncComponent} from 'components/async_load';
...
+    bottom: number;
+    left?: number;
+    right: number;
+}
 
+export interface Props {
+    style?: PickerStyle;
+    rightOffset: number;
+    topOffset: number;
+    leftOffset: number;
```
Comment (M-ZubairAhmed) [inline components/emoji_picker/emoji_picker_tabs.tsx:28]: even these are optional, that why they have defaults

**Review by M-ZubairAhmed — CHANGES_REQUESTED** [review]

File: components/emoji_picker/emoji_picker_overlay.tsx [hunk 3]
```diff
@@ -10,12 +10,12 @@ import {Constants} from 'utils/constants';
...
 
-import EmojiPickerTabs from './emoji_picker_tabs.jsx';
+import EmojiPickerTabs from './emoji_picker_tabs';
 
 type Props = {
     show: boolean;
-    container?: () => ReactNode | ReactNode;
-    target: () => ReactNode | ReactNode;
+    container?: () => ReactNode;
+    target: () => ReactNode;
```
Comment (sinansonmez) [inline components/emoji_picker/emoji_picker_overlay.tsx:18]: Thanks for cleaning them out

**Review by sinansonmez — CHANGES_REQUESTED** [review]
@alexpjohnson great work. I just have one tiny comment

File: components/emoji_picker/emoji_picker_tabs.tsx [hunk 4]
```diff
@@ -11,29 +10,42 @@ import {makeAsyncComponent} from 'components/async_load';
...
+    placement?: ('top' | 'bottom' | 'left' | 'right');
+    customEmojis?: Emoji[];
```
[the remark this answers is not in the document]
Comment (M-ZubairAhmed) [reply components/emoji_picker/emoji_picker_tabs.tsx:30]: yes its not used and can be safely removed. cc @alexpjohnson

File: components/emoji_picker/emoji_picker_tabs.tsx [hunk 5]
```diff
@@ -11,29 +10,43 @@ import {makeAsyncComponent} from 'components/async_load';
...
-    };
+type PickerStyle = {
+    top: number;
+    bottom: number;
+    left?: number;
+    right?: number;
+}
+
+export interface Props {
+    style?: PickerStyle;
```
Comment (M-ZubairAhmed) [inline components/emoji_picker/emoji_picker_tabs.tsx:25]: this style is actually `React.CSSProperties` if you see 
https://github.com/mattermost/mattermost-webapp/pull/11401/files#diff-fc703a1770091015a25d4016640ec6e6a3531ecda84d59a915f9af5f55aaed9cL76
its being passed down to `Tabs` as style object

**Review by M-ZubairAhmed — CHANGES_REQUESTED** [review]

**Review by M-ZubairAhmed — APPROVED** [review]
💯 It looks good to me. Thank you 🎉

**Review by sinansonmez — CHANGES_REQUESTED** [review]
Thanks for the changes, just one last comment to improve readability IMHO

File: components/emoji_picker/emoji_picker_tabs.tsx [hunk 6]
```diff
@@ -70,16 +75,22 @@ export default class EmojiPickerTabs extends PureComponent {
...
             }
 
-            pickerStyle.top = pickerStyle.top ? pickerStyle.top + this.props.topOffset : this.props.topOffset;
+            let offset = 0;
+            if (pickerStyle.top) {
+                offset = (this.props.topOffset || 0) + (pickerStyle.top as number);
+            } else {
+                offset = (this.props.topOffset || 0);
+            }
+            pickerStyle.top = offset;
```
Comment (sinansonmez) [inline components/emoji_picker/emoji_picker_tabs.tsx:90]: I think we can improve this part
```suggestion
            pickerStyle.top = this.props.topOffset || 0;
            if (pickerStyle.top && this.props.topOffset) {
                pickerStyle.top = this.props.topOffset + (pickerStyle.top as number);
            }
```

File: components/emoji_picker/emoji_picker_tabs.tsx [hunk 7]
```diff
@@ -11,29 +10,35 @@ import {makeAsyncComponent} from 'components/async_load';
...
-        leftOffset: PropTypes.number,
-        placement: PropTypes.oneOf(['top', 'bottom', 'left', 'right']),
-        customEmojis: PropTypes.object,
-        onEmojiClose: PropTypes.func.isRequired,
-        onEmojiClick: PropTypes.func.isRequired,
-        onGifClick: PropTypes.func,
-        enableGifPicker: PropTypes.bool,
-    };
+export interface Props {
+    style?: React.CSSProperties;
```
Comment (M-ZubairAhmed) [inline components/emoji_picker/emoji_picker_tabs.tsx:18]: sorry minor nit : can you please also import this selectively

**Comment by jgilliam17** [comment]
Thanks @alexpjohnson 
There is an issue with the emoji picker opening at the top of the screen (see test server screenshot) 
Screen Shot 2022-11-07 at 7 35 09 PM

File: components/emoji_picker/emoji_picker_tabs.tsx [hunk 8]
```diff
@@ -70,16 +75,19 @@ export default class EmojiPickerTabs extends PureComponent {
...
-            pickerStyle.top = pickerStyle.top ? pickerStyle.top + this.props.topOffset : this.props.topOffset;
+            pickerStyle.top = this.props.topOffset || 0;
+            if (pickerStyle.top && this.props.topOffset) {
+                pickerStyle.top = this.props.topOffset + (pickerStyle.top as number);
+            }
 
-            if (pickerStyle.left) {
-                pickerStyle.left += this.props.leftOffset;
+            if (pickerStyle.left && this.props.leftOffset) {
+                (pickerStyle.left as number) += this.props.leftOffset;
```
Comment (M-ZubairAhmed) [inline components/emoji_picker/emoji_picker_tabs.tsx:90]: @alexpjohnson can we take a look at this logic one more time, it seems that it is causing the issue of positioning mentioned above.
Fallback to that would be to just revert this piece of logic we are not able to solve the issue

**Review by M-ZubairAhmed — CHANGES_REQUESTED** [review]

File: components/emoji_picker/emoji_picker_tabs.tsx (hunk 8 above)
Comment (alexpjohnson) [reply components/emoji_picker/emoji_picker_tabs.tsx]: Yup. It was a problem with the initial condition of the if statement. Should be good now

**Review by M-ZubairAhmed — APPROVED** [review]

**Comment by M-ZubairAhmed** [comment]
@alexpjohnson can you please resolve the conflicts

**Comment by M-ZubairAhmed** [comment]
@jgilliam17 can you please check this one out :)

**Comment by jgilliam17** [comment]
@alexpjohnson can you please rebase branch with master so I can re-trigger e2es. Thanks 🙂

**Comment by alexpjohnson** [comment]
> @alexpjohnson can you please rebase branch with master so I can re-trigger e2es. Thanks slightly_smiling_face

Done!

**Review by jgilliam17 — APPROVED** [review]
Thank you @alexpjohnson 
Tested, looks good to merge.
[E2E report](https://mattermost-cypress-report.s3.amazonaws.com/275178-afbad04-onprem-ent-webapp-pr-11401-mm-ee-test:afbad04/mochawesome.html) - no PR related failures 
@M-ZubairAhmed can you please approve the last workflow and merge? Thanks 🙂

**Pull request merged**